src/System.Text.Json/src/System/Text/Json/Node/JsonString.cs

```diff
@@ -235,43 +239,50 @@ internal bool TryGetBytesFromBase64(out byte[] value)
             int bufferSize = _value.Length / 4 * 3;
 
             byte[] arrayToReturnToPool = null;
-            try
-            {
-                Span<byte> buffer = bufferSize <= JsonConstants.StackallocThreshold
-                    ? stackalloc byte[bufferSize]
-                    : arrayToReturnToPool = ArrayPool<byte>.Shared.Rent(bufferSize);
+            Span<byte> buffer = bufferSize <= JsonConstants.StackallocThreshold
+                ? stackalloc byte[bufferSize]
+                : arrayToReturnToPool = ArrayPool<byte>.Shared.Rent(bufferSize);
 
-                if (Convert.TryFromBase64String(_value, buffer, out int bytesWritten))
-                {
-                    value = buffer.Slice(0, bytesWritten).ToArray();
-                    return true;
-                }
-                else
-                {
-                    value = default;
-                    return false;
-                }
+            if (Convert.TryFromBase64String(_value, buffer, out int bytesWritten))
+            {
+                buffer = buffer.Slice(0, bytesWritten);
+                value = buffer.ToArray();
+                buffer.Clear();
+                ReturnToPool(arrayToReturnToPool);
+                return true;
             }
-            finally
+            else
             {
-                if (arrayToReturnToPool != null)
-                {
-                    Array.Clear(arrayToReturnToPool, 0, arrayToReturnToPool.Length);
-                    ArrayPool<byte>.Shared.Return(arrayToReturnToPool);
-                }
+                value = default;
+                buffer.Clear();
+                ReturnToPool(arrayToReturnToPool);
+                return false;
             }
+
 #else
             try
             {
                 value = Convert.FromBase64String(_value);
                 return true;
             }
-            catch
+            catch (FormatException)
             {
                 value = null;
                 return false;
             }
 #endif
         }
+
+        /// <summary>
+        /// Clears a byte array and returns it to the pool.
```

Comment:
It does not clear, it just returns to the pool 😉 

Make the method as
```c#
private void ClearAndReturnToPool(byte[] arrayToReturnToPool, int bufferSize)
{
    if (arrayToReturnToPool != null)
    {
        arrayToReturnToPool.AsSpan(0, bufferSize).Clear();
        ArrayPool<byte>.Shared.Return(arrayToReturnToPool);
    }
}
```
so the call-sites can be simplified.

Reply:
> It does not clear, it just returns to the pool 😉

I totally missed this when making changes back and forth. Nice catch!

So, to be clear (pun intended), the `arrayToReturnToPool` and not `buffer` need to be cleared before returning it to the pool? I'm confused considering [this comment](https://github.com/dotnet/corefx/pull/41467#discussion_r332219294). Does the current solution of clearing `buffer` not clear the data that is returned to the pool?

Reply:
* when renting from the pool `arrayToReturnToPool` may be bigger in size than `bufferSize`
* `ArrayPool<T>.Return` has an overload that allows for clearing, but it clears the whole array
* as only `bufferSize` data is actually used, there is no need to clear the whole array --> it's enough to clear just the subset of actually used data, which is `bufferSize`
* so to don't do any unnecessary work just clear the slice of used data => `arrayToReturnToPool.AsSpan(0, bufferSize).Clear()`

Regarding your question `buffer` is the span-represenation of `arrayToReturnToPool`. Which one gets cleared doesn't matter, as they "point" to the same data.

Is it clear what needs to be cleared?

Reply:
> So, to be clear (pun intended), the `arrayToReturnToPool` and not `buffer` need to be cleared before returning it to the pool? I'm confused considering [this comment](https://github.com/dotnet/corefx/pull/41467#discussion_r332219294). Does the current solution of clearing `buffer` not clear the data that is returned to the pool?

+1 to what @gfoidl said.

Clearing `buffer` does clear the data.
Clearing either will do it (`buffer` is a view into the contents of the ` arrayToReturnToPool`).

Reply:
@gfoidl - why pass in `buffersize` to `ClearAndReturnToPool` and not the `buffer` span itself? We already have to slice it to the correct amount.

Reply:
I can't give an answer :wink: 
You are right, passing `buffer` itself would be better.

Reply:
Alright, then it is as I expected. Thanks for the input.